### DISCUSSION TITLE
Is it possible to make Canvas renderer for GTK4 using Gio?

### DISCUSSION THREAD

**Discussion by Pazystamas** [tdewolff/canvas#228] [Ideas]
Hello,
Is there a way to make GTK4 renderer for Canvas (adapt gio renderer) ? Autogenerated GTK4 bindings for Go https://github.com/diamondburned/gotk4 uses gio, so probably it should work somehow? Should it use gtk snapshot?

**Comment by tdewolff** [comment]
We can build a GTK renderer, but note that [GTK/Gio](https://docs.gtk.org/gio/) is something else than [GioUI](https://gioui.org/). I suppose there are various ways in connecting Canvas with GTK, perhaps drawing to a PNG and rendering that, or directly to a pixel buffer, or perhaps using Cairo or using OpenGL (Wayland). I would need to investigate what is most performant, since either we invoke the rasterizer in Canvas (slow), or use the rasterizer of Cairo (most likely to be better optimized), or use OpenGL but that incurs the cost of converting all paths to triangles first, sending it to the GPU, drawing, and sending it back to the CPU. I'm pretty sure OpenGL would be by far the slowest option.
